### ISSUE TITLE
[Need help] Heroku Always Crashing.

### ISSUE THREAD

**Issue by glennposadas** [firebase/flashlight#108]
### Commendation

First of all, I was amazed by this project when I saw the demo for search. I became a fan of Firebase the first time I discovered its features, as a mobile developer with no background in server side programming. With that being said, all the configurations of this flashlight for my own project has been complex to me. Let me state all of the steps I made and the questions I wanna ask. I will try to format this post as neat as I can.

### My future contribution
If I ever succeed, with your help of course, in setting all of the stuff needed for the Firebase+ElasticSearch, I will make a tutorial step by step for newbies like me.

### EDIT:
I have removed my long post that contains all the stupid steps that I made.

### TL;DR:
Just follow carefully the guide of this repo. That's all! You'll need to install several CLIs and softwares. 💃 Now it's working. Damn! I got my own working project in my local. 

### More questions:

1. Can two or more users use my app if I only have sandbox/free account on Heroku?
2. Everytime I change something in my project files, I commit, push, and heroku restart. Is this correct?
3. Last important question, to test if everything works fine, I should just click ```OPEN APP``` on my Heroku project page, right? Or access the file ```index.html```?
4. Finally, whatever I do, the log says: 
```
2016-12-15T17:17:27.274344+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/" host=radio-one-flashlight.herokuapp.com request_id=4a9dc6b1-cb6f-4b26-801d-5ba49c811303 fwd="49.144.202.234" dyno= connect= service= status=503 bytes=
```

**Comment by katowulf** [comment]
> 1. Can two or more users use my app if I only have sandbox/free account on Heroku?

Not sure what this means. Any number of clients should be able to make requests. You only need one service worker deployed to Heroku to serve those requests. 

> 2. Everytime I change something in my project files, I commit, push, and heroku restart. Is this correct?

Yes, once you are working in production. But you should test locally (more below).

> 3. Last important question, to test if everything works fine, I should just click OPEN APP on my Heroku project page, right? Or access the file index.html?

Sort of. You can run the index.html locally on your dev machine or deploy it somewhere. Keep in mind that the example/ folder represents your **client** whereas the app.js and lib/ folder represent the server process answering the requests.

Thus, you need your `example/example.js` to match the database URL given in your `config.js` file. Additionally, unless you are also using `messages` and `users` as your data types, you'll need to tweak the `type` of your searches to match your real database structure. 

All of this is probably best done running a local ElasticSearch instance and pointing everything to localhost:9200 for testing.  Once you have that working,  I'd try deploying things to Heroku and using your production config settings.

> 4. Finally, whatever I do, the log says: [error message]
There has to be some more logging available that explains what happened with the app; a stack trace showing the real error. 

Alternately, if you tried to create a web worker rather than a service worker on Heroku, it might just be looking to see if anything is answering HTTP requests, which this script wouldn't, of course.

☼, Kato

**Comment by glennposadas** [comment]
thanks, Kato. I will try to comprehend everything you've said. But right now, I need to access my index.html to test everything. If you have time, can you please check my "About the Service-Account" in this post? I really believe my setup was just wrong that's why the whole project crashes. I just don't like having a test first in my local, because my mindset is like "what's the use of testing something in my local if I can't deploy that something online properly?".

**Comment by katowulf** [comment]
The use of local testing is to isolate the moving parts and test them separately so you can determine which one is failing. `service-account.json` needs to be in the same path as your app.js file. Unless you've modified it, it should just work.

**Comment by glennposadas** [comment]
I'm currently doing all of your steps 1 by 1 again. This time, I installed node.js in my local and there's now node modules in my project. Perhaps this is the step that I needed. :D I'm so excited to see this working for my own firebase project. I'll get back here later. It's 3:50AM here. Thanks again!

**Comment by glennposadas** [comment]
hi @katowulf It's freakin working now! I just had to follow your steps carefully and google some things. One last question before we close this:
Where to go from here? I mean I need to of course implement all of this in the client side (e.g. my mobile app project or POSTMAN or whatever client side software and call POST and GET http requests).

**Comment by katowulf** [comment]
Much simpler to let Flashlight manage the ES interactions than to try POST/GET and what-not. So essentially follow the examples in `example/example.js` and write/read from the `search/` path in Firebase. As long as the Flashlight service is running, it will handle the ElasticSearch interactions and put the results out in JSON format to Firebase.

**Comment by katowulf** [comment]
Also, congrats! 🔥🔥🔥

**Comment by glennposadas** [comment]
Thanks again, Kato. I got what you said. So I'll just rewrite your example.js into my mobile app platform. :)

**Comment by glennposadas** [comment]
hi @katowulf , I was planning to just post questions on SO cause I thought you might have had enough of noob questions from me lol. Anyways, I've been trying to figure out what you have told me. Ditch my plan in making POST/GET (rest api) and instead do something like your ```example.js```. What do you mean by that? I really couldn't figure out how should I connect my mobile app to this. Of course the app is already connected to Firebase, but I need this elasticsearch.

P.S. Javascript can be used in iOS. This is a Javascript right (nodejs)? Thanks!

**Comment by katowulf** [comment]
You should think of the Flashlight service as your _back end_, and your mobile client (be it iOS, JavaScript, or whatnot) as the _client_.  Flashlight handles all the interactions with ES--you don't need to (but I guess you could if the data is all world readable) query it directly via a REST API.

You use Firebase on the client, but implementing on of the SDKs and writing to the `query` path (this is what is illustrated in example.js). If you want to do this in iOS, just use the iOS SDK to write to that path and listen for results.
